File: go/pkg/dns/dns_test.go
package dns

import (
	"strings"
	"testing"

	"github.com/ThirdKeyAi/schemapin/go/pkg/crypto"
	"github.com/ThirdKeyAi/schemapin/go/pkg/discovery"
)

// --- ParseTxtRecord tests ---

func TestParseFullRecord(t *testing.T) {
	r, err := ParseTxtRecord("v=schemapin1; kid=acme-2026-01; fp=sha256:abcd1234")
	if err != nil {
		t.Fatal(err)
	}
	if r.Version != "schemapin1" {
		t.Errorf("expected version 'schemapin1', got %q", r.Version)
	}
	if r.Kid != "acme-2026-01" {
		t.Errorf("expected kid 'acme-2026-01', got %q", r.Kid)
	}
	if r.Fingerprint != "sha256:abcd1234" {
		t.Errorf("expected fingerprint 'sha256:abcd1234', got %q", r.Fingerprint)
	}
}

func TestParseMinimalRecord(t *testing.T) {
	r, err := ParseTxtRecord("v=schemapin1;fp=sha256:abc")
	if err != nil {
		t.Fatal(err)
	}
	if r.Version != "schemapin1" {
		t.Errorf("expected version 'schemapin1', got %q", r.Version)
	}
	if r.Kid != "" {
		t.Errorf("expected empty kid, got %q", r.Kid)
	}
	if r.Fingerprint != "sha256:abc" {
		t.Errorf("expected fingerprint 'sha256:abc', got %q", r.Fingerprint)
	}
}

func TestParseLowercasesFingerprint(t *testing.T) {
	r, err := ParseTxtRecord("v=schemapin1; fp=SHA256:ABCDEF")
	if err != nil {
		t.Fatal(err)
	}
	if r.Fingerprint != "sha256:abcdef" {
		t.Errorf("expected fingerprint 'sha256:abcdef', got %q", r.Fingerprint)
	}
}

func TestParseToleratesWhitespaceAndOrder(t *testing.T) {
	r, err := ParseTxtRecord("  fp = sha256:beef ;  v = schemapin1  ")
	if err != nil {
		t.Fatal(err)
	}
	if r.Version != "schemapin1" {
		t.Errorf("expected version 'schemapin1', got %q", r.Version)
	}
	if r.Fingerprint != "sha256:beef" {
		t.Errorf("expected fingerprint 'sha256:beef', got %q", r.Fingerprint)
	}
}

func TestParseIgnoresUnknownFields(t *testing.T) {
	r, err := ParseTxtRecord("v=schemapin1; fp=sha256:abc; future=ignoreme; another=value")
	if err != nil {
		t.Fatal(err)
	}
	if r.Fingerprint != "sha256:abc" {
		t.Errorf("expected fingerprint 'sha256:abc', got %q", r.Fingerprint)
	}
}

func TestParseMissingVFails(t *testing.T) {
	_, err := ParseTxtRecord("fp=sha256:abc")
	if err == nil {
		t.Fatal("expected error for missing 'v' field")
	}
	if !strings.Contains(err.Error(), "'v'") {
		t.Errorf("error should mention 'v'; got: %v", err)
	}
}

func TestParseMissingFpFails(t *testing.T) {
	_, err := ParseTxtRecord("v=schemapin1")
	if err == nil {
		t.Fatal("expected error for missing 'fp' field")
	}
	if !strings.Contains(err.Error(), "'fp'") {
		t.Errorf("error should mention 'fp'; got: %v", err)
	}
}

func TestParseUnsupportedVersionFails(t *testing.T) {
	_, err := ParseTxtRecord("v=schemapin99; fp=sha256:abc")
	if err == nil {
		t.Fatal("expected error for unsupported version")
	}
	if !strings.Contains(err.Error(), "unsupported version") {
		t.Errorf("error should mention 'unsupported version'; got: %v", err)
	}
}

func TestParseFpWithoutSha256PrefixFails(t *testing.T) {
	_, err := ParseTxtRecord("v=schemapin1; fp=abc")
	if err == nil {
		t.Fatal("expected error for fp without sha256: prefix")
	}
	if !strings.Contains(err.Error(), "sha256:") {
		t.Errorf("error should mention 'sha256:' prefix; got: %v", err)
	}
}

func TestParseFieldWithoutEqualsFails(t *testing.T) {
	_, err := ParseTxtRecord("v=schemapin1; broken")
	if err == nil {
		t.Fatal("expected error for field without '='")
	}
	if !strings.Contains(err.Error(), "'='") {
		t.Errorf("error should mention '='; got: %v", err)
	}
}

// --- VerifyDnsMatch tests ---

func makeKeypair(t *testing.T) (string, string) {
	t.Helper()
	km := crypto.NewKeyManager()
	priv, err := km.GenerateKeypair()
	if err != nil {
		t.Fatal(err)
	}
	privPEM, err := km.ExportPrivateKeyPEM(priv)
	if err != nil {
		t.Fatal(err)
	}
	pubPEM, err := km.ExportPublicKeyPEM(&priv.PublicKey)
	if err != nil {
		t.Fatal(err)
	}
	return privPEM, pubPEM
}

func TestVerifyMatch(t *testing.T) {
	_, pubPEM := makeKeypair(t)
	km := crypto.NewKeyManager()
	fp, err := km.CalculateKeyFingerprintFromPEM(pubPEM)
	if err != nil {
		t.Fatal(err)
	}

	disc := &discovery.WellKnownResponse{
		SchemaVersion: "1.4",
		PublicKeyPEM:  pubPEM,
		DeveloperName: "Test Dev",
	}
	txt := &DnsTxtRecord{
		Version:     "schemapin1",
		Fingerprint: strings.ToLower(fp),
	}
	if err := VerifyDnsMatch(disc, txt); err != nil {
		t.Errorf("expected match, got error: %v", err)
	}
}

func TestVerifyMismatch(t *testing.T) {
	_, pubPEM := makeKeypair(t)
	disc := &discovery.WellKnownResponse{
		SchemaVersion: "1.4",
		PublicKeyPEM:  pubPEM,
		DeveloperName: "Test Dev",
	}
	txt := &DnsTxtRecord{
		Version:     "schemapin1",
		Fingerprint: "sha256:0000000000000000000000000000000000000000000000000000000000000000",
	}
	err := VerifyDnsMatch(disc, txt)
	if err == nil {
		t.Fatal("expected mismatch error")
	}
	if !strings.Contains(err.Error(), "mismatch") {
		t.Errorf("error should mention 'mismatch'; got: %v", err)
	}
}

// --- TxtRecordName tests ---

func TestTxtRecordName(t *testing.T) {
	cases := []struct {
		in, want string
	}{
		{"example.com", "_schemapin.example.com"},
		{"example.com.", "_schemapin.example.com"},
		{"sub.example.com", "_schemapin.sub.example.com"},
	}
	for _, c := range cases {
		got := TxtRecordName(c.in)
		if got != c.want {
			t.Errorf("TxtRecordName(%q): got %q, want %q", c.in, got, c.want)
		}
	}
}

File: javascript/tests/dns.test.js
/**
 * Tests for DNS TXT cross-verification (v1.4).
 */

import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { KeyManager } from '../src/crypto.js';
import { parseTxtRecord, verifyDnsMatch, txtRecordName } from '../src/dns.js';

// ---------------------------------------------------------------------------
// parseTxtRecord
// ---------------------------------------------------------------------------

describe('parseTxtRecord', () => {
    it('should parse a full record', () => {
        const r = parseTxtRecord('v=schemapin1; kid=acme-2026-01; fp=sha256:abcd1234');
        assert.equal(r.version, 'schemapin1');
        assert.equal(r.kid, 'acme-2026-01');
        assert.equal(r.fingerprint, 'sha256:abcd1234');
    });

    it('should parse a minimal record (no kid)', () => {
        const r = parseTxtRecord('v=schemapin1;fp=sha256:abc');
        assert.equal(r.version, 'schemapin1');
        assert.equal(r.kid, null);
        assert.equal(r.fingerprint, 'sha256:abc');
    });

    it('should lowercase the fingerprint', () => {
        const r = parseTxtRecord('v=schemapin1; fp=SHA256:ABCDEF');
        assert.equal(r.fingerprint, 'sha256:abcdef');
    });

    it('should tolerate whitespace and field order', () => {
        const r = parseTxtRecord('  fp = sha256:beef ;  v = schemapin1  ');
        assert.equal(r.version, 'schemapin1');
        assert.equal(r.fingerprint, 'sha256:beef');
    });

    it('should ignore unknown fields (forward-compat)', () => {
        const r = parseTxtRecord('v=schemapin1; fp=sha256:abc; future=ignoreme');
        assert.equal(r.fingerprint, 'sha256:abc');
        assert.equal(r.version, 'schemapin1');
    });

    it('should reject when v is missing', () => {
        assert.throws(() => parseTxtRecord('fp=sha256:abc'), /missing required 'v'/);
    });

    it('should reject when fp is missing', () => {
        assert.throws(() => parseTxtRecord('v=schemapin1'), /missing required 'fp'/);
    });

    it('should reject unsupported version', () => {
        assert.throws(
            () => parseTxtRecord('v=schemapin99; fp=sha256:abc'),
            /unsupported version/
        );
    });

    it('should reject fp without sha256: prefix', () => {
        assert.throws(
            () => parseTxtRecord('v=schemapin1; fp=abc'),
            /must be sha256/
        );
    });

    it('should reject a field that has no = sign', () => {
        assert.throws(
            () => parseTxtRecord('v=schemapin1; broken'),
            /missing '='/
        );
    });

    it('should tolerate a trailing semicolon', () => {
        const r = parseTxtRecord('v=schemapin1; fp=sha256:abc;');
        assert.equal(r.fingerprint, 'sha256:abc');
    });
});

// ---------------------------------------------------------------------------
// txtRecordName
// ---------------------------------------------------------------------------

describe('txtRecordName', () => {
    it('should prefix _schemapin', () => {
        assert.equal(txtRecordName('example.com'), '_schemapin.example.com');
    });

    it('should strip a trailing dot from the domain', () => {
        assert.equal(txtRecordName('example.com.'), '_schemapin.example.com');
    });
});

// ---------------------------------------------------------------------------
// verifyDnsMatch
// ---------------------------------------------------------------------------

describe('verifyDnsMatch', () => {
    function makeDiscovery(publicKeyPem) {
        return {
            schema_version: '1.4',
            developer_name: 'Dev',
            public_key_pem: publicKeyPem
        };
    }

    it('should accept a matching fingerprint', () => {
        const { publicKey } = KeyManager.generateKeypair();
        const fp = KeyManager.calculateKeyFingerprint(publicKey).toLowerCase();
        const txt = { version: 'schemapin1', kid: null, fingerprint: fp };
        // Should not throw
        verifyDnsMatch(makeDiscovery(publicKey), txt);
    });

    it('should accept a matching fingerprint case-insensitively', () => {
        const { publicKey } = KeyManager.generateKeypair();
        const fp = KeyManager.calculateKeyFingerprint(publicKey).toUpperCase();
        // Mimic parser output: parser lowercases everything.
        const txt = { version: 'schemapin1', kid: null, fingerprint: fp.toLowerCase() };
        verifyDnsMatch(makeDiscovery(publicKey), txt);
    });

    it('should throw on fingerprint mismatch', () => {
        const { publicKey } = KeyManager.generateKeypair();
        const txt = {
            version: 'schemapin1',
            kid: null,
            fingerprint: 'sha256:0000000000000000000000000000000000000000000000000000000000000000'
        };
        assert.throws(() => verifyDnsMatch(makeDiscovery(publicKey), txt), /mismatch/);
    });

    it('should throw when discovery is missing public_key_pem', () => {
        const txt = {
            version: 'schemapin1',
            kid: null,
            fingerprint: 'sha256:abc'
        };
        assert.throws(() => verifyDnsMatch({}, txt), /missing public_key_pem/);
    });
});
